service/raft/raft_group0.cc

```diff
@@ -420,14 +420,22 @@ future<> raft_group0::join_group0(std::vector<gms::inet_address> seeds, bool as_
         if (server == nullptr) {
             // This is the first time discovery is run. Create and start a Raft server for group 0 on this node.
             raft::configuration initial_configuration;
+            bool nontrivial_snapshot = false;
             if (g0_info.id == my_id) {
                 // We were chosen as the discovery leader.
                 // We should start a new group with this node as voter.
                 group0_log.info("Server {} chosen as discovery leader; bootstrapping group 0 from scratch", my_id);
                 initial_configuration.current.emplace(my_addr, true);
+                // Force snapshot transfer from us to subsequently joining servers.
+                // This is important for upgrade and recovery, where the group 0 state machine
+                // (schema tables in particular) is nonempty.
+                // In a fresh cluster this will trigger an empty snapshot transfer which is redundant but correct.
```

Comment:
I don't think the snapshot cluster in a fresh cluster will be "empty". It will contain mutations from system tables, which are not empty. This was exactly my point during the call. Previously, Scylla fresh cluster setup was not transferring any data. This was important for example when you create a fresh cluster using *heterogeneous* versions. In this case each node would have its own initial state of the system tables. After this change, the leader of the cluster will be pushing its system tables state over to all nodes, regardless of their version. It could be harmless, but it is not asked for.

Reply:
> I don't think the snapshot in a fresh cluster will be "empty". It will contain mutations from system tables, which are not empty.

Right, I'll fix the comment.

> It could be harmless

It **must** be harmless, if not - we have a bug.

Reply:
> Right, I'll fix the comment.

Actually, it's wrong... The snapshot will be empty.

What do you mean by "state of system tables"? I thought you're talking about schema and topology tables. If those are nonempty, then Raft log is nonempty - we modify them by committing commands after all.

We don't send any internal state like `system.local`. That would be unsafe of course.

So, in a fresh cluster the snapshot **will** be empty but the Raft log won't be empty because of initially created distributed tables.

> Imagine a fresh cluster containing two nodes, 5.2 and 5.3. 5.3 snapshot will contain 5.3 mutations. They will be sent over to 5.2 node is 5.3 is the leader. These mutations are not guaranteed to be harmless on 5.2.

In the same way any commands committed by the 5.3 node may not be safe. 

This is an orthogonal problem to snapshotting. It's the problem of cluster features and implementing a correct upgrade procedure in general. A 5.2 node should not be able to join 5.3 cluster if there's an incompatible feature.

Reply:
You're right that there is no practical difference today and the situation I described should be handled by the feature compatibility check.

I still cringe when I reason about this code, however, and here's why.

This patch establishes and documents a new contract - which was until now implicit:
- snapshot index 0 means the snapshot is empty and Raft leader doesn't need to transfer a snapshot
- snapshot index 1 means the snapshot is non-empty and Raft leader needs to transfer a snapshot to every new follower.

And then you misuse this contract in this very patch. A new node has an empty raft snapshot, and we used snapshot index 0 for it until now. Yet after this patch, whether or not a node is fresh, we will use snapshot index 1. Of course transferring an empty snasphot is harmless (in theory). But if it is harmless, what use case are we supporting by allowing snapshot index 0? Maybe we should require snapshot index 1 in all cases?

I think the use of index 1 in the code blurs the very protocol the code establishes.

Reply:
Given the patch is fixing a CI issue, perhaps it should be approved. But perhaps I at least was able to make my point clear.

Reply:
> snapshot index 1 means the snapshot is non-empty and Raft leader needs to transfer a snapshot to every new follower.

> And then you misuse this contract in this very patch.

None of the comments said "snapshot index 1 means that the snapshot is non-empty". The comments are only saying that snapshot index 1 will enforce snapshot transfer, which is **necessary** when your snapshot is non-empty, but it doesn't imply that your snapshot is non-empty. And we require nothing about empty snapshots in this case, it's fine to transfer them, it's fine not to transfer them.

Reply:
> None of the comments said "snapshot index 1 means that the snapshot is non-empty".

In fact it can be empty since the command may have been "dummy".

Reply:
Alright, but what point does it make to allow snapshot index 0 then? We ourselves are not using it.

Reply:
We are setting snapshot index to 1 only on the first server in group 0. Second, third etc. must have it set to 0. Otherwise Raft would assume that log is in sync and not send a snapshot.

Reply:
But for the first server indeed there is no point to ever set it to 0. Maybe as a micro-optimization if you know that your snapshot is empty, but not worth the brain cycles.

Reply:
// Note: there are two sensible options for the initial snapshot
    // used by the first server in a Raft cluster:
    // - if the initial snapshot's index is 0, Raft will assume that
    //   the state machine is empty,
    // - if the initial snapshot's index is 1, Raft will have to perform
    //   a snapshot transfer before replicating commands to any server
    //   which joins with empty Raft log (in particular with  snapshot index 0).
    // Thus, if you want to create a Raft cluster with a non-empty state
    // machine, so that joining servers always receive a snapshot,
    // you should set the initial snapshot index on the first server to 1
    // and on all subsequently joining servers to 0.
    
    
    You say here specifically: there is no point in having snapshot index 0 for the leader. 
    You say in the comment: there are two sensible options for the first server. 
    Looks like option of index 0 is not sensible after all?

Reply:
Ideally, let's ban snapshot index 0 for the first server. I understand detecting the first server on raft level is impossible. But then again, what is our stock suggestion? After all, for raft groups which don't require a pre-defined initialization order, it makes sense to set snapshot index 0 on all servers? The implicitness of the contract is what frustrates me.

Reply:
> You say in the comment: there are two sensible options for the first server. 
Looks like option of index 0 is not sensible after all?

It is sensible as a micro-optimization when you have an empty snapshot.

> But then again, what is our stock suggestion? After all, for raft groups which don't require a pre-defined initialization order, it makes sense to set snapshot index 0 on all servers? The implicitness of the contract is what frustrates me.

The stock suggestion would be:
- for all servers present in the initial configuration, set snapshot index to 1. (And make sure that they have the same initial state machine state.)
- for all other servers (i.e. those which will join through a configuration change), set snapshot index to 0.